Parser is looking in the wrong place for the tsconfig file

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

I apologize for the size of this [repro](https://github.com/Tbhesswebber/typescript-eslint-debug) - I am in a weird spot where I am using (Narwhal Extensions)[nx.dev] with [nx-electron](https://www.npmjs.com/package/nx-electron) and somehow the whole shebang has managed to cause the parser to lose track of the base tsconfig in _some_ circumstances.

To run linting on every piece of the project, you have to run `nx run-many --target lint --all --parallel` (the parallel flag just speeds it all up).

This is a super greenfield POC that just got started, so it's essentially empty, but I'd be happy to jump into a chat to better understand/explain the problem in order to file a better issue with whomever the real culprit is if it's not the parser.

```sh
# error message being received (file paths are condensed)
...galvanize/apps/g-ent-e2e/src/integration/app.spec.ts
  0:0  error  Parsing error: Cannot read file '.../galvanize/node_modules/tsconfig.base.json'
```

Running with typescript-eslint in debug mode hasn't seemed to provide any clarity.

**Expected Result**

I expected the parser to look at the root of the workspace for `tsconfig.base.json`

**Actual Result**

A number of files are causing parsing errors at line 0 column 0 because the parser is looking for `./node_modules/tsconfig.base.json` rather than `./tsconfig.base.json`.  This does not happen with all of the files.

**Additional Info**

There are a couple hundred more lines of output from running `DEBUG=typescript-eslint:* nx run-many --target lint --all --parallel`, but none seem to be concerning.

```sh
typescript-eslint:typescript-estree:createProjectProgram Creating project program for: /Users/tannerhesswebber/Documents/code/galvanize/galvanize/apps/g-ent-desktop/src/environments/environment.prod.ts +330ms
  typescript-eslint:typescript-estree:createWatchProgram Found existing program for file. /users/tannerhesswebber/documents/code/galvanize/galvanize/apps/g-ent-desktop/src/environments/environment.prod.ts +330ms
  typescript-eslint:parser:parser Resolved libs from program: [ 'es2017', 'dom' ] +329ms

/Users/tannerhesswebber/Documents/code/galvanize/galvanize/apps/g-ent-desktop/src/environments/environment.ts
  0:0  error  Parsing error: Cannot read file '/users/tannerhesswebber/documents/code/galvanize/galvanize/node_modules/tsconfig.base.json'
```

**Versions**

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/parser` | `4.3.0` |
| `TypeScript`                | `4.0.3` |
| `ESLint`                    | `7.10.0` |
| `node` (system)                      | `12.16.3` |
| `electron` | `7.1.0`|

--- a/packages/typescript-estree/src/create-program/createDefaultProgram.ts
+++ b/packages/typescript-estree/src/create-program/createDefaultProgram.ts
@@ -4,7 +4,7 @@ import * as ts from 'typescript';
 import { Extra } from '../parser-options';
 import {
   ASTAndProgram,
-  getTsconfigPath,
+  CanonicalPath,
   createDefaultCompilerOptionsFromExtra,
 } from './shared';
 
@@ -27,7 +27,7 @@ function createDefaultProgram(
     return undefined;
   }
 
-  const tsconfigPath = getTsconfigPath(extra.projects[0], extra);
+  const tsconfigPath: CanonicalPath = extra.projects[0];
 
   const commandLine = ts.getParsedCommandLineOfConfigFile(
     tsconfigPath,

--- a/packages/typescript-estree/src/create-program/createWatchProgram.ts
+++ b/packages/typescript-estree/src/create-program/createWatchProgram.ts
@@ -9,7 +9,6 @@ import {
   CanonicalPath,
   createDefaultCompilerOptionsFromExtra,
   getCanonicalFileName,
-  getTsconfigPath,
 } from './shared';
 
 const log = debug('typescript-eslint:typescript-estree:createWatchProgram');
@@ -197,9 +196,7 @@ function getProgramsForProjects(
    * - the required program hasn't been created yet, or
    * - the file is new/renamed, and the program hasn't been updated.
    */
-  for (const rawTsconfigPath of extra.projects) {
-    const tsconfigPath = getTsconfigPath(rawTsconfigPath, extra);
-
+  for (const tsconfigPath of extra.projects) {
     const existingWatch = knownWatchProgramMap.get(tsconfigPath);
 
     if (existingWatch) {

--- a/packages/typescript-estree/src/create-program/shared.ts
+++ b/packages/typescript-estree/src/create-program/shared.ts
@@ -60,10 +60,6 @@ function ensureAbsolutePath(p: string, extra: Extra): string {
     : path.join(extra.tsconfigRootDir || process.cwd(), p);
 }
 
-function getTsconfigPath(tsconfigPath: string, extra: Extra): CanonicalPath {
-  return getCanonicalFileName(ensureAbsolutePath(tsconfigPath, extra));
-}
-
 function canonicalDirname(p: CanonicalPath): CanonicalPath {
   return path.dirname(p) as CanonicalPath;
 }
@@ -105,5 +101,4 @@ export {
   ensureAbsolutePath,
   getCanonicalFileName,
   getScriptKind,
-  getTsconfigPath,
 };

--- a/packages/typescript-estree/src/parser-options.ts
+++ b/packages/typescript-estree/src/parser-options.ts
@@ -1,6 +1,7 @@
 import { DebugLevel } from '@typescript-eslint/types';
 import { Program } from 'typescript';
 import { TSESTree, TSNode, TSESTreeToTSNode, TSToken } from './ts-estree';
+import { CanonicalPath } from './create-program/shared';
 
 type DebugModule = 'typescript-eslint' | 'eslint' | 'typescript';
 
@@ -19,7 +20,7 @@ export interface Extra {
   loc: boolean;
   log: (message: string) => void;
   preserveNodeMaps?: boolean;
-  projects: string[];
+  projects: CanonicalPath[];
   range: boolean;
   strict: boolean;
   tokens: null | TSESTree.Token[];

--- a/packages/typescript-estree/src/parser.ts
+++ b/packages/typescript-estree/src/parser.ts
@@ -12,7 +12,12 @@ import { createSourceFile } from './create-program/createSourceFile';
 import { Extra, TSESTreeOptions, ParserServices } from './parser-options';
 import { getFirstSemanticOrSyntacticError } from './semantic-or-syntactic-errors';
 import { TSESTree } from './ts-estree';
-import { ASTAndProgram, ensureAbsolutePath } from './create-program/shared';
+import {
+  ASTAndProgram,
+  CanonicalPath,
+  ensureAbsolutePath,
+  getCanonicalFileName,
+} from './create-program/shared';
 
 const log = debug('typescript-eslint:typescript-estree:parser');
 
@@ -109,46 +114,53 @@ function resetExtra(): void {
   };
 }
 
+function getTsconfigPath(tsconfigPath: string, extra: Extra): CanonicalPath {
+  return getCanonicalFileName(ensureAbsolutePath(tsconfigPath, extra));
+}
+
 /**
- * Normalizes, sanitizes, resolves and filters the provided
+ * Normalizes, sanitizes, resolves and filters the provided project paths
  */
 function prepareAndTransformProjects(
   projectsInput: string | string[] | undefined,
   ignoreListInput: string[],
-): string[] {
-  let projects: string[] = [];
+): CanonicalPath[] {
+  const sanitizedProjects: string[] = [];
 
   // Normalize and sanitize the project paths
   if (typeof projectsInput === 'string') {
-    projects.push(projectsInput);
+    sanitizedProjects.push(projectsInput);
   } else if (Array.isArray(projectsInput)) {
     for (const project of projectsInput) {
       if (typeof project === 'string') {
-        projects.push(project);
+        sanitizedProjects.push(project);
       }
     }
   }
 
-  if (projects.length === 0) {
-    return projects;
+  if (sanitizedProjects.length === 0) {
+    return [];
   }
 
   // Transform glob patterns into paths
-  const globbedProjects = projects.filter(project => isGlob(project));
-  projects = projects
-    .filter(project => !isGlob(project))
-    .concat(
-      globSync([...globbedProjects, ...ignoreListInput], {
-        cwd: extra.tsconfigRootDir,
-      }),
-    );
+  const nonGlobProjects = sanitizedProjects.filter(project => !isGlob(project));
+  const globProjects = sanitizedProjects.filter(project => isGlob(project));
+  const uniqueCanonicalProjectPaths = new Set(
+    nonGlobProjects
+      .concat(
+        globSync([...globProjects, ...ignoreListInput], {
+          cwd: extra.tsconfigRootDir,
+        }),
+      )
+      .map(project => getTsconfigPath(project, extra)),
+  );
 
   log(
     'parserOptions.project (excluding ignored) matched projects: %s',
-    projects,
+    uniqueCanonicalProjectPaths,
   );
 
-  return projects;
+  return Array.from(uniqueCanonicalProjectPaths);
 }
 
 function applyParserOptionsToExtra(options: TSESTreeOptions): void {
@@ -252,8 +264,9 @@ function applyParserOptionsToExtra(options: TSESTreeOptions): void {
   // NOTE - ensureAbsolutePath relies upon having the correct tsconfigRootDir in extra
   extra.filePath = ensureAbsolutePath(extra.filePath, extra);
 
-  // NOTE - prepareAndTransformProjects relies upon having the correct tsconfigRootDir in extra
-  const projectFolderIgnoreList = (options.projectFolderIgnoreList ?? [])
+  const projectFolderIgnoreList = (
+    options.projectFolderIgnoreList ?? ['**/node_modules/**']
+  )
     .reduce<string[]>((acc, folder) => {
       if (typeof folder === 'string') {
         acc.push(folder);
@@ -262,6 +275,7 @@ function applyParserOptionsToExtra(options: TSESTreeOptions): void {
     }, [])
     // prefix with a ! for not match glob
     .map(folder => (folder.startsWith('!') ? folder : `!${folder}`));
+  // NOTE - prepareAndTransformProjects relies upon having the correct tsconfigRootDir in extra
   extra.projects = prepareAndTransformProjects(
     options.project,
     projectFolderIgnoreList,

